### ISSUE TITLE
Search result was not shown when I search by uppercase keyword.

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1072]
## Phenominon

When I tried to search with uppercase keyword, sometimes I couldn't find any result even if the result had existed for searching of lowercase keyword.
## Bug position

I fixed  this bug by changing "searchtools.js_t" located in "C:\Python27\Lib\site-packages\Sphinx-1.1.3-py2.7.egg\sphinx\themes\basic\static"
after installation of python-2.7.3.msi & setuptools-0.6c11.win32-py2.7.exe.

**searchtools.js_t**

```
#!javascript

- 55            this.performSearch(query);
+ 55           this.performSearch(query.toLowerCase());
```

Please consider to commit this fixation in the near future.

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1072
- Originally reported by: Donghyun Lee
- Originally created at: 2013-01-10T12:54:02.097

**Comment by shimizukawa** [comment]
_From Jon Waltman on 2013-01-11 04:10:13+00:00_

Some people might consider that a feature.  E.g., you want to distinguish between a capitalized class name and a lowercase function name.

I couldn't find an example where case affected the results.

**Comment by shimizukawa** [comment]
_From Donghyun Lee on 2013-01-11 06:32:31+00:00_

As an example, I attached the file, html_changed.zip which I made for the manual of CUBRID which is open source DBMS made on my company in Korea.
(This manual is just a sample which I'm writing now. so it's incomplete.)

As I understand, Sphinx search doesn't distinguish between capitalized keyword and lowercase keyword.

You can find the phenomenon that I described above if you try to search the keyword "SQL_LOG_MAX_SIZE" using html_original.zip.

html_changed.zip is the manual which I changed the searching script by using above method.
1. searching "SQL_LOG_MAX_SIZE" using html_original.zip
   => you can't find any result.
2. searching "sql_log_max_size" using html_original.zip
   => you can find 2 results, which are written with "SQL_LOG_MAX_SIZE" on the text.
3. searching "SQL_LOG_MAX_SIZE" using html_changed.zip
   => you can find 2 results, which are written with "SQL_LOG_MAX_SIZE" on the text.

You can see that _static/searchtools.js file is changed from the original one.

**Comment by shimizukawa** [comment]
_From Jon Waltman on 2013-01-11 08:19:25+00:00_

Thanks for the example.

This is indeed an issue: an uppercase search for an uppercase word fails while the lowercase search finds the uppercase word.

The issue is a result of "stemming" uppercase words differently than lowercase words.

I'm not sure if just lowercasing everything is the best approach though since case should be a factor in ranking the results.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2013-01-11 08:29:26+00:00_

Hmm, I thought we already did lower-case query terms... (at least for fulltext search), but it makes sense also for object names.

**Comment by shimizukawa** [comment]
_From Richard Sternagel on 2013-06-04 20:18:03+00:00_

I have a similar issue due to two current facts:
- **every** term which is added to the index is lower cased (see [en](https://github.com/sphinx-doc/sphinx/blob/2c1cbf319e4af29993cc495252ed060cde652c98/sphinx/search/en.py#L238) & [ja](https://github.com/sphinx-doc/sphinx/blob/2c1cbf319e4af29993cc495252ed060cde652c98/sphinx/search/ja.py#L276))
- the result of stemmed upper case words **differs** from lower case words in the js code (don't know where this happens but [this is the line it then matters](https://github.com/sphinx-doc/sphinx/blob/2c1cbf319e4af29993cc495252ed060cde652c98/sphinx/themes/basic/static/searchtools.js_t#L159))

My issue word is `FILENAME`/`filename`:

So during index creation `FILENAME` and `filename` both become `filenam` (first lower the string, then stem it, then add it). But when I later on search via my browser within my generated documentation `filename` is stemmed into `filenam` and `FILENAME` is stemmed into `filename` which obviously doesn't exist in the index. This is because of this line in `searchtools.js_t` (same as link above):

```
#!javascript

var word = stemmer.stemWord(tmp[i]).toLowerCase();
```

... which should imo be:

```
#!javascript

var word = stemmer.stemWord(tmp[i].toLowerCase());
```

Again, **first** lower it, **then** stem it and then look it up. This would not be needed if (the js) stemming wouldn't give different results for lower/upper case. I'm not sure why this happens and if this is a bug or a feature.

[Jon Waltman](https://bitbucket.org/jonwaltman):

> I'm not sure if just lowercasing everything is the best approach though since case should be a factor in ranking the results.

I agree, but at the moment no uppercase word accomplishes to get in the (terms) index anyway.

**Comment by shimizukawa** [comment]
_From Pascal Boulerie on 2013-10-03 11:42:07+00:00_

A specific issue about uppercase search sensitivity in Japanese is #1111 and has been fixed : 1111 #1111

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-01-12 20:29:17+00:00_

Closes #1072: In the JS search, fix issues searching for upper-cased words by lowercasing words before stemming.

→ <<cset b1936aa1f3e8dd9f0547aff67f09849da788d08e>>
